test/libsolidity/SolidityABIJSON.cpp

```diff
@@ -292,6 +308,8 @@ BOOST_AUTO_TEST_CASE(events)
 	{
 		"name": "f",
 		"constant": false,
+		"view": false,
```

Comment:
Did we consider a variant where state-changes are an attribute of the function but memory-changes are attributes of the function arguments? I guess creating new objects in newly allocated memory should be fine for functions that do not alter memory, because I think altering memory should essentially not be about memory as a whole, but only about pre-existing objects in memory.
So we can just apply a constant keyword to function arguments of reference type (it does not make sense for return values) and for complex variables, this constant keyword is pulled all the way through.
If we do it like that, is there a way to get more fine-grained access in the future, i.e. you can pass an array where it is fine for the function to add new elements but it is not allowed to alter existing elements? We could add a 'mutable' keyword that negates the effect of 'constant' starting from that level.

First iteration allows:
```
function f(uint[][] constant x) { ... }
```
Second iteration allows:
```
function f(uint[]constant[] x) { ... }
```